File: 2017/day9/day9.py
#!/usr/bin/env python3
# [Day 9: Stream Processing](https://adventofcode.com/2017/day/9)

import sys
from pathlib import Path

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()


def compute(data, part1=True):
    in_group = 0
    in_garbage = False
    score = 0
    garbage = 0

    i = 0
    while i < len(data):
        c = data[i]

        if c == "!":
            i += 2
            continue

        if in_garbage:
            if c == ">":
                in_garbage = False
            else:
                garbage += 1
            i += 1
            continue

        if c == "<":
            in_garbage = True
            i += 1
            continue

        if c == "{":
            in_group += 1

        elif c == "}":
            score += in_group
            in_group -= 1
            assert in_group >= 0

        i += 1

    return score if part1 else garbage


assert compute("{}") == 1
assert compute("{{{}}}") == 6
assert compute("{{},{}}") == 5
assert compute("{{{},{},{{}}}}") == 16
assert compute("{<a>,<a>,<a>,<a>}") == 1
assert compute("{{<ab>},{<ab>},{<ab>},{<ab>}}") == 9
assert compute("{{<!!>},{<!!>},{<!!>},{<!!>}}") == 9
assert compute("{{<a!>},{<a!>},{<a!>},{<ab>}}") == 3

assert compute("<>", False) == 0
assert compute("<random characters>", False) == 17
assert compute("<<<<>", False) == 3
assert compute("<{!>}>", False) == 2
assert compute("<!!>", False) == 0
assert compute("<!!!>>", False) == 0
assert compute("""<{o"i!a,<{i<a>""", False) == 10

print(compute(data, True))
print(compute(data, False))

File: 2017/day9/day9.rs
//! [Day 9: Stream Processing](https://adventofcode.com/2017/day/9)

/// Initialize from the puzzle input.
#[must_use]
pub fn solve(data: &str) -> (u32, u32) {
    let mut it = data.trim_ascii().bytes();

    let mut in_group = 0;
    let mut in_garbage = false;
    let mut score = 0;
    let mut garbage = 0;

    while let Some(c) = it.next() {
        if c == b'!' {
            it.next();
        } else if in_garbage {
            if c == b'>' {
                in_garbage = false;
            } else {
                garbage += 1;
            }
        } else if c == b'<' {
            in_garbage = true;
        } else if c == b'{' {
            in_group += 1;
        } else if c == b'}' {
            score += in_group;
            in_group -= 1;
        }
    }

    (score, garbage)
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn part1() {
        assert_eq!(solve("{}").0, 1);
        assert_eq!(solve("{{{}}}").0, 6);
        assert_eq!(solve("{{},{}}").0, 5);
        assert_eq!(solve("{{{},{},{{}}}}").0, 16);
        assert_eq!(solve("{<a>,<a>,<a>,<a>}").0, 1);
        assert_eq!(solve("{{<ab>},{<ab>},{<ab>},{<ab>}}").0, 9);
        assert_eq!(solve("{{<!!>},{<!!>},{<!!>},{<!!>}}").0, 9);
        assert_eq!(solve("{{<a!>},{<a!>},{<a!>},{<ab>}}").0, 3);
    }

    #[test]
    fn part2() {
        assert_eq!(solve("<>").1, 0);
        assert_eq!(solve("<random characters>").1, 17);
        assert_eq!(solve("<<<<>").1, 3);
        assert_eq!(solve("<{!>}>").1, 2);
        assert_eq!(solve("<!!>").1, 0);
        assert_eq!(solve("<!!!>>").1, 0);
        assert_eq!(solve(r#"<{o"i!a,<{i<a>"#).1, 10);
    }
}
